### ISSUE TITLE
The class ReturnTypeWillChange doesn’t have the Attribute attribute

### ISSUE THREAD

**Issue by derrabus** [vimeo/psalm#6849]
When I run Psalm 4.12.0 on a codebase that uses the new `ReturnTypeWillChange` attribute, I get a lot of errors regarding that attribute. The message is always:

> The class ReturnTypeWillChange doesn’t have the Attribute attribute

Interestingly enough, this only happens when I run Psalm on PHP 8.1 (8.1.0-rc5 currently). If I use PHP 8.0.12 to run Psalm on the very same codebase, the error is not reported.

When using reflection to analyze `ReturnTypeWillChange` on PHP 8.1, we can see that the class actually has that attribute: https://3v4l.org/H7ngu

**Comment by derrabus** [comment]
Thank you for your quick reply, dear bot. No, the problem is not reproducible on psalm.dev. I assume that psalm.dev uses PHP 8.0 or older to perform the analysis.

**Comment by orklah** [comment]
8.0 must have been fixed by https://github.com/vimeo/psalm/pull/6732

I'll look at what happens on 8.1

**Comment by orklah** [comment]
Do you have an example of failing code? I'm trying to reproduce on 8.1 on CI but so far, no luck

**Comment by derrabus** [comment]
I noticed this issue when bumping Psalm for the `doctrine/dbal` repository. On that repository, we suppress `UndefinedAttributeClass` errors for the `ReturnTypeWillChange` class. This is why I didn't notice that Psalm just issues different errors for both PHP versions.

I've created a small reproducer: https://github.com/derrabus/psalm-return-type-will-change

Run:

```sh
composer install
./vendor/bin/psalm --no-cache
```

Here's the output I get.

#### PHP 8.0.12
```
ERROR: UndefinedAttributeClass - src/EmptyCollection.php:11:7 - Attribute class ReturnTypeWillChange does not exist (see https://psalm.dev/241)
    #[ReturnTypeWillChange]
```

#### PHP 8.1.0 RC5
```
ERROR: InvalidAttribute - src/EmptyCollection.php:11:7 - The class ReturnTypeWillChange doesn’t have the Attribute attribute (see https://psalm.dev/242)
    #[ReturnTypeWillChange]
```

**Comment by orklah** [comment]
Ok, I get it.

When you run psalm without a version specified, it uses the minimum version of PHP supported by your composer file, so in this case 8.0.

This means you're running with PHP 8.1 but Psalm will run as if it was checking PHP 8.0 code, so it doesn't include the needed stub.

Can you try to add --php-version=8.1 and confirm it solves your issue?

**Comment by derrabus** [comment]
Indeed, that does fix the issue, thank you very much for the very quick investigation. However, I'm not sure if this is the right solution. The error message I get on PHP 8.1 _without raising the project's language level_ is still highly confusing.

Attributes are designed to be used even if they're not defined. This allows us to use an attribute that PHP 8.1 introduces in a forward-compatible manner, even if the language level of the codebase is PHP 8.0 or even lower. In this very case, I would even argue that we should not emit an error no matter what the language level is.

That being said, the error message I currently get on PHP 8.0 is acceptable to me and I can easily ignore it with an issue handler:

```xml
<UndefinedAttributeClass>
    <errorLevel type="suppress">
        <referencedClass name="ReturnTypeWillChange"/>
    </errorLevel>
</UndefinedAttributeClass>
```

However, the error message I get when running Psalm with PHP 8.1 is plain wrong. My understanding is that I should either get the same error message (because I'm using a PHP 8.1 class on a codebase with language level 8.0) or none at all (because the `ReturnTypeWillChange` class is defined correctly by the PHP runtime).

**Comment by orklah** [comment]
Note: I created https://github.com/vimeo/psalm/issues/6854 to discuss about the particular case of ReturnTypeWillChange. It's important we make the difference between this attribute that has been made to be on version that don't support it and other attributes.

> Attributes are designed to be used even if they're not defined

I have to think more about it but except for ReturnTypeWillChange, I think Psalm *should* report cases where an undefined attribute is used. The potential for errors and typo is just too great if we allow undefined symbols pass through

> My understanding is that I should either get the same error message (because I'm using a PHP 8.1 class on a codebase with language level 8.0) or none at all (because the ReturnTypeWillChange class is defined correctly by the PHP runtime).

You're right, I'll have to do more tests to see why PHP 8.1 running with --php-version=8.0 manage to see the attribute but without seeing the `Attribute` attribute

**Comment by derrabus** [comment]
> > Attributes are designed to be used even if they're not defined
> 
> I have to think more about it but except for ReturnTypeWillChange, I think Psalm _should_ report cases where an undefined attribute is used. The potential for errors and typo is just too great if we allow undefined symbols pass through

I agree 100%!

**Comment by derrabus** [comment]
To add a bit of complexity to the problem: if I add the composer package `symfony/polyfill-php81` to the project which brings a polyfill for the attribute, the error on PHP 8.0 is gone. But on PHP 8.1, Psalm still complains about the missing `Attribute` attribute.

**Comment by orklah** [comment]
Oh, that's not fun to fix...

So here's what's going on. The `ReturnTypeWillChange` attribute is indeed found by Psalm as a ClassLike. Unfortunately, there is no attribute registered on it when we don't load the PHP 8.1 stubs (for example when we're analyzing PHP 8.0 from a PHP 8.1 environment like you).

It seemed pretty simple so I did exactly that: using reflection in order to retrieve the attributes from an internal class.

Unfortunately for me, running Psalm on itself would get me errors saying `ReflectionClass::getAttributes` returns mixed. And I realized that Psalm still supports PHP 7.1, where `ReflectionClass::getAttributes` was not a thing at all.

So I can't use Reflection in older versions because it's unavailable and it becomes useless on higher versions because stubs kicks in.

I'm quite lucky because we stated in #6854 that we'll make Psalm ignore this particular attribute, but otherwise, I don't really have a solution (except using something like BetterReflection which is waaay above my current skills with Psalm)

**Comment by TimoLehnertz** [comment]
I am getting the same error regardless of if I add `--php-version=8.0` or 8.1.
The Error only shows up when using PHP version 8.1. It's not showing when using PHP 8.0.
Psalm is on version 5.x-dev.
The attribute is indeed missing in [CoreGenericClasses.phpstub](https://github.com/vimeo/psalm/blob/4.x/stubs/CoreGenericClasses.phpstub)

**Comment by orklah** [comment]
It's there on 4.x in the file you linked: https://github.com/vimeo/psalm/blob/4.x/stubs/CoreGenericClasses.phpstub#L499

And it's also there on master: https://github.com/vimeo/psalm/blob/master/stubs/CoreGenericClasses.phpstub#L499

Are you sure you're using the latest commit on 5.x?

**Comment by TimoLehnertz** [comment]
You're right, it's there indeed. When I run psalm --version, I get `Psalm 5.x-dev@`. I took a look in my psalm file and the attribute shows up in there as well. No idea what could cause psalm to not find it.
